File: 17 Dynamic programming/MinAbsSum/minAbsSum.py
def solution(A):
    N = len(A)
    M = 0

    for i in range(N):
        A[i] = abs(A[i])
        M = max(A[i], M)

    S = sum(A)
    count = [0] * (M + 1)

    for i in range(N):
        count[A[i]] += 1

    dp = [-1] * (S + 1)
    dp[0] = 0

    for a in range(1, M + 1):
        if count[a] > 0:
            for j in range(S):
                if dp[j] >= 0:
                    dp[j] = count[a]
                elif j >= a and dp[j - a] > 0:
                    dp[j] = dp[j - a] - 1

    result = S

    for i in range(S // 2 + 1):
        if dp[i] >= 0:
            result = min(result, S - 2 * i)
    return result


''' 
O(N * max(abs(A))**2)
A = [1, 5, 2, -2]
print(solution(A)) '''

File: Lessons/17 Dynamic programming/MinAbsSum/minAbsSum.js
function solution(A) {
  if (A.length == 0) { return 0; }
  let minDiff = Infinity;

  A.forEach((val, index) => A[index] = Math.abs(val));
  A.sort((a, b) => a - b);

  const max = Math.max(...A);

  const count = Array(max + 1).fill(0);

  for (let i = 0; i < A.length; i += 1) {
    count[A[i]] += 1;
  }

  const total = A.reduce((a, b) => a + b, 0);

  const dp = Array(total).fill(-1);
  dp[0] = 0;

  for (let i = 0; i < count.length; i += 1) {
    if (count[i] > 0) {
      for (let j = 0; j < dp.length; j += 1) {
        if (dp[j] >= 0) {
          dp[j] = count[i];
        } else if (j >= i && dp[j - i] > 0) {
          dp[j] = dp[j - i] - 1;
        }
        if (dp[j] >= 0) {
          if (j == (total / 2)) { return 0; }
          minDiff = Math.min(minDiff, Math.abs(total - 2 * j));
        }
      }
    }
  }
  return minDiff;
}
// console.log(solution([1, 5, 2, -2]));
